Modules/Filtering/FFT/include/itkFFTWFFTImageFilterFactory.h

```diff
@@ -0,0 +1,70 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkFFTWFFTImageFilterFactory_h
+#define itkFFTWFFTImageFilterFactory_h
+#include "ITKFFTExport.h"
+
+#include "itkObjectFactoryBase.h"
+
+namespace itk
+{
+/**
+ *\class FFTWFFTImageFilterFactory
+ * \brief Create instances of FFTW FFT ImageFilter objects using an object factory.
+ * \ingroup ITKFFT
+ */
+class ITKFFT_EXPORT FFTWFFTImageFilterFactory : public ObjectFactoryBase
```

Comment:
@thewtex Could you expand on why `FFTWFFTImageFilterFactory` and `VnlFFTImageFilterFactory` are necessary for these changes? It looks like these classes are replacing the individualized default factory registration methods we've previously set up, but are still registering `FFTImageFilterFactory` objects for FFT image filter overrides to the object factory singleton list. 

I'm not clear on the advantage of changing the current setup other than readability for grouping Vnl or FFTW overrides together. At the same time, `FFTWFFTImageFilterFactory` and `VnlFFTImageFilterFactory` never directly override filter instantiations and I'm not sure they should be called factories at all.

Reply:
`itkFFTWFFTImageFilterFactory` and `itkVnlFFTImageFilterFactory` are called in Python to register the factories. The class names come from the factory initialization convention for C++, also implemented in all the IO classes. The pattern is:

`itk<impl><factory_type>Factory`

Where `<impl>`, `<factory_type>` come from the *itk-module.cmake* `FACTORY_NAMES`, which provides `<factory_type>::<impl>`.

Without the grouping, we would need 6 `Factory` classes in C++, 6 wrapped classes per Vnl, FFTW, VkFFT implementation, which does seem unnecessary. Do you agree?

>  At the same time, FFTWFFTImageFilterFactory and VnlFFTImageFilterFactory never directly override filter instantiations and I'm not sure they should be called factories at all.

They do -- otherwise the tests would not pass :-)

Reply:
Let me try to rephrase.

When we look at [`itk::ObjectFactoryBase`](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkObjectFactoryBase.h#L42) we see that its purpose is to instantiate objects:

> ObjectFactoryBase is used to create itk objects. The base class
  ObjectFactoryBase contains a static method CreateInstance() that is
  used to create itk objects from the list of registerd ObjectFactoryBase
  sub-classes

This is the primary advantage of a "Factory" class in ITK, it can take in a type name and instantiate a different (subclass) type in its place via the `CreateInstance` method. Classes inheriting from `itk::ObjectFactoryBase` define narrow instantiation overrides that, when they are registered with the global singleton list of object factories, can be used to swap out construction procedures at runtime. 

`FFTWFFTImageFilterFactory` and `VnlFFTImageFilterFactory` as they are defined muddy the waters a bit on what purpose a "factory" has.

Properties that `FFTWFFTImageFilterFactory` and `VnlFFTImageFilterFactory` share with factories:
- They inherit from `itk::ObjectFactoryBase`
- They function to create override behavior so that FFT image filters are instantiated with a specific backend

Properties that `FFTWFFTImageFilterFactory` and `VnlFFTImageFilterFactory` do not share with factories:
- They cause overrides to be registered as a side effect of their construction and have no direct mechanism for removing/disabling overrides
- They are never added to the ITK list of object factories that are consulted when `CreateInstance` is called
- Their own `CreateObject` method is never called
- They call another factory, `FFTImageFilterFactory`, to do the actual work of handling overrides

The disadvantage of naming these as "factories" is that it obscures the dependency on `FFTImageFilterFactory` and could be misleading, since all other factories need to be registered whereas these two do not.

I think it's a matter of preference whether FFT classes should be registered together or in separately templated factories. I do want to point out that the IO factory convention refers only to a single, non-templated class at a time, and factories are always one-to-one with the type of file to read. As we expand factory registration procedures it would be nice to standardize what "unit" relationship we can expect between factories and classes that they override. `FFTImageFilterFactory` walks this line by accepting partial templates, so an individual factory object maintains a one-to-one relationship between a superclass and a subclass while in practice overriding approximately eight template specializations over different image input/outputs.

How do you feel about the following suggestions?
- If you'd prefer to stick with grouping Vnl/FFTW FFT override initializations into one class, these could be renamed something like `FFTWFFTImageFilterFactoryInitializer` and `VnlFFTImageFilterFactoryInitializer` to more distinctly clarify their relationship to `FFTImageFilterFactory` and each group of FFT implementation class templates. There may be CMake issues to resolve with name changes.
- Otherwise, something closer to the original method of defining individual factory registrations for Vnl and FFTW subclasses may help better preserve clarity and object factory conventions.

Sorry for the essay. I will add that I agree your implementation as it currently stands would be practical and allow us to close #2233. My point is more to what conventions we should aim to establish as we expand object factory functionality.

Reply:
@thewtex Would it be reasonable to split off FFT-specific factory changes to a separate PR to allow Python changes to go forward?

I'm convinced that satisfactorily answering the question of how to register groups of factories together (such as FFT factories) in a future-proof manner will require more extensive CMake changes than necessary to address the question of Python factory initialization here. Existing CMake architecture is well-suited to classes with _simple_ usage defined _across modules_, such as ImageIO factories, that have very limited base class variations and can be encompassed in a single [configuration file](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CMake/itkImageIOFactoryRegisterManager.h.in) corresponding to their base class. This architecture implies that we should only aim to default register factories for any base class that has an accompanying, dedicated `itk<base_class>FactoryRegisterManager.h.in` file in ITK.

But, what if we want to group classes such as FFT image filter registrations together? By grouping FFT filters in a similar [configuration file](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CMake/itkFFTImageFilterFactoryRegisterManager.h.in) we've already blurred some lines because there is no common "FFTImageFilter" base object to reference, but instead eight independent classes. What if we want use a different mechanism such as a `VnlFFTImageFilterFactoryInitializer` to register a set of class backends? Or accompany those registrations with other, arbitrary factories that will optimize operations based on the first registrations, such as overriding `DiscreteGaussianImageFilter` with `FFTDiscreteGaussianImageFilter` for GPU-accelerated FFT blurring? Our existing CMake architecture is not flexible enough to answer these questions.

I propose that we pursue the following:
- For this PR and most likely for the ITK 5.3 release, leave the eight Vnl and FFTW FFT factories explicitly spelled out in `itk-module.cmake` and remove changes to initializations. (May still be necessary to rename factory prefixes to get CMake running correctly, i.e. `FFT::Forward1D` -> `FFTImageFilter::Forward1D`)
- In a subsequent PR for ITK 5.3 or 5.4, explore the following changes:
  - Add parsing for a `INIT_DEPENDS` or similarly-named field (`PYTHON_DEPENDS`, `FACTORY_DEPENDS`, etc) in `itk-module.cmake` to inform Python of which parent modules should load the given dependent module so that initializations are performed correctly. Note that this is similar to the current PR, but extends beyond explicitly defined classes such as ImageIO.
  - Add a CMake mechanism in `itk_module_impl` to configure arbitrary module initialization code. This is where factory grouping code such as a `VnlFFTImageFilterFactoryInitializer` could run with similar static initialization as currently done in [`itkFFTImageFilterFactoryRegisterManager`](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CMake/itkFFTImageFilterFactoryRegisterManager.h.in), but without being tied to a particular cross-module base class.

@thewtex feedback on this idea? If you think it is sound I would be willing to take this on in early March, or assist if anyone else wants to take a stab at it earlier.

Reply:
@tbirdso The existing FFT factory initialization has bugs in its functionality, related to not using the existing CMake architecture properly, and is more complicated than it needs to be. This should be addressed in 5.3. Also, I do not think we need another CMake mechanism or arbitrary module initialization -- this is more complicated than it needs to be. I understand the desire for consistent conventions because this can help understanding, but the conventions exist to provide functionality, functionality does not try to work its way into existence for functionality's sake. If we need to clarify or broaden our understanding of conventions, then we should do so. The eight initializations are not functional in this case -- in practice for FFT's you almost always want to initialize all filters, especially the forward and backward pairs because they are likely not compatible across implementations. And, from an implementation and build perspective, it is unnecessarily burdensome and verbose. Maybe we could use an `Initializer` class as suggested to help the implementation, I will look into the details.

Reply:
> The existing FFT factory initialization has bugs in its functionality, related to not using the existing CMake architecture properly, and is more complicated than it needs to be. This should be addressed in 5.3.

Could you please clarify the CMake bugs that these changes are intended to address? I'm unclear on if these are functional issues, in which case I absolutely agree they should be addressed in 5.3, or conventions (i.e. using `FFT` instead of `FFTImageFilter`), in which case I'm still inclined to agree but less certain of their urgency.

> Also, I do not think we need another CMake mechanism or arbitrary module initialization -- this is more complicated than it needs to be.

That's fine, but it still leaves open the question of how to initialize arbitrary factories. Existing CMake architecture provides no path for a module to initialize factories that are not of an enumerated type (`ImageIO`, `TransformIO`, `MeshIO`, or `FFTImageFilter`).

> I understand the desire for consistent conventions because this can help understanding, but the conventions exist to provide functionality, functionality does not try to work its way into existence for functionality's sake. If we need to clarify or broaden our understanding of conventions, then we should do so.

Generally agreed, but there are other paths forward here than violating conventions and redefining class functionality. 

> The eight initializations are not functional in this case -- in practice for FFT's you almost always want to initialize all filters, especially the forward and backward pairs because they are likely not compatible across implementations. And, from an implementation and build perspective, it is unnecessarily burdensome and verbose.

I agree that the burden can be reduced, however, I'm not sure I fully understand the cost-benefit tradeoff here. To me it seems that the benefit is a reduction of a handful of repetitive CMake lines while the cost is obscuring object factory intent.

> Maybe we could use an `Initializer` class as suggested to help the implementation, I will look into the details.

👍 Thanks @thewtex . If this is an urgent fix feel free to override me here and I can open an issue for future discussion on solidifying conventions.

Reply:
> Could you please clarify the CMake bugs that these changes are intended to address? 

Here I am referring to the special-casing in the CMake and factory initialization for the FFT factory. This special-casing is not necessary and it does not work with additional functionality that we need.

C++ and CMake changes in this patch are not independent of the Python additions in this patch. While making the Python additions, I found inconsistencies with the FFT factories and the CMake configuration, which should be made independently but are also required for the Python functionality. We need the factory mechanisms:

- To work with CMake module configuration (ideally in a simple way)
- To work with Python, ideally in a simple way
- To work with C++ binary static initialization
- To work with Python run-time invocation 

> Existing CMake architecture provides no path for a module to initialize factories that are not of an enumerated type

It can and without extra complexity and infrastructure.

> Generally agreed, but there are other paths forward here than violating conventions and redefining class functionality.

There is still work to do on this patch, but impressions of conventions may need to be refined.

>  To me it seems that the benefit is a reduction of a handful of repetitive CMake lines while the cost is obscuring object factory intent.

There is no intent to obsure object factory intent. The cost is not just a handful of repetitive CMake lines -- it is multiple classes, wrapping for those classes, infrastructure to use those classes, CMake module complexity, and, most importantly "client" side complexity to use factories when they are needed.

Thanks for the thoughts and comments, but please be patient and allow me to follow-up with code changes and  information that should help clarify.

Reply:
From discussion with @thewtex:
- This patch is intended to resolve Slicer issues discussed in [https://discourse.slicer.org/t/addressing-issue-with-pip-install-itk/21763/2](https://discourse.slicer.org/t/addressing-issue-with-pip-install-itk/21763/2). The advantages of the fix outweigh the potential for any technical debt we may incur from defining "InitFactory" classes that are not themselves factories.
- We anticipate that "InitFactory" will be the primary vehicle for initialization of arbitrary factories at Python runtime going forward. This naming scheme is currently required for ITK Python to parse factories and create factory initialization lists.
- It is reasonable to have "InitFactory" classes inherit from `itk::LightObject` rather than `itk::ObjectFactoryBase`. I will add that change to this PR.
- I will investigate and help resolve lingering wrapping failures holding up this PR.